sprint-3

alxgrents opened the pull request.

alxgrents committed:
sprint_3 Добавил роуты и роутер

alxgrents committed:
sprint_3 Доработал роуты, начал интеграцию с swagger api

alxgrents committed:
sprint_3 Добавил тесты, почти внедрил интеграцию с swagger api но он упал

alxgrents committed:
sprint_3 Добавил работу с Api, начал добавлять работу с Socket-api

alxgrents committed:
sprint_3 [wip] Добавил работу с сокетом

alxgrents committed:
sprint_3 Добавил работу с сокетом

agadoev commented on package.json:
```diff
@@ -3,12 +3,20 @@
+    "jsdom": "^20.0.0",
+    "less": "4.1.3",
+    "mocha": "^10.0.0",
+    "ts-node": "^10.9.1"
```
Все эти зависимости можно смело помещать в devDependencies

agadoev commented on package.json:
```diff
@@ -3,12 +3,20 @@
+    "@types/chai": "^4.3.3",
+    "@types/jsdom": "^20.0.0",
+    "@types/mocha": "^9.1.1",
+    "@types/sinon": "^10.0.13",
```
Давай использовать статические версии зависимостей

agadoev commented on src/common/base-block.ts:
```diff
@@ -13,6 +12,7 @@ import {
+import {createProxy} from "../utils/create-proxy";
 
 enum Events {
     init = 'init',
```
Подумай над тем, чтобы использовать вместо enuma'a использовать discriminated union. Фишка в том, что при транспиляции tyepscript кода enum превращается в объект, который будет жить в рантайме, а union'ы - нет. Так же сейчас определение enum'a выглядит избыточно, потому что ты по два раза по сути пишешь названия события, что не круто.
В общем, если можно воспользоваться union'ом, лучше пользоваться union'ом.
[тут](https://stackoverflow.com/questions/40275832/typescript-has-unions-so-are-enums-redundant) подробнее объяснено

agadoev commented on src/components/form/Form.ts:
```diff
@@ -1,54 +1,74 @@
         );
+        this._labels = {};
+        Object.keys(this._inputs).forEach((name) => {
+            const label = content.querySelector(`label[for=${name}]`) as HTMLLabelElement;
```
Если можно обойтись без явного приведения типов, лучше обойтись без явного приведения типов. Этот механизм на самом деле довольно опасный и костыльный. Тут можно написать так:
```suggestion
            const label = content.querySelector<HTMLLabelElement>(`label[for=${name}]`);
```

agadoev commented on src/components/form/Form.ts:
```diff
@@ -1,54 +1,74 @@
-type FormOptions = BaseBlockOptions & {
+export type FormOptions = BaseBlockOptions & {
     validators?: Record<string, Validator>
+    onSubmit?: (data: Record<string, any>) => any,
```
может лучше использовать void?
```suggestion
    onSubmit?: (data: Record<string, any>) => void,
```

agadoev commented on src/components/chat-nav-item/ChatNavItem.ts:
```diff
@@ -1,14 +1,38 @@
+import {BaseBlockOptions, BaseBlockProps} from "../../common/types";
+import {ChatData} from "../../data";
+
+type ChatNavItemOptions  = BaseBlockOptions & {
+    chat: ChatData;
+}
 
 class ChatNavItem extends BaseBlock {
+    protected _props: BaseBlockProps & {
+        chat: ChatData,
+    }
```
Сейчас с первого взгляда не очень понятно чем отличаются Options и Props и зачем было сделано это разделение. Так же не совсем понятно зачем объединять типы Options и Props дочернего компонента с базовым, они по логике являются разными сущностями.
Я бы предложил сделать класс Block параметризуемым с помощью generic'ов, чтобы можно было писать так:
```suggestion
type ChatNavItemOptions  = BaseBlockOptions & {
    chat: ChatData;
}

class ChatNavItem extends BaseBlock<ChatNavItemProps> {
    protected _props: BaseBlockProps & {
        chat: ChatData,
    }
```
В таком случае по идее дочерний компонент уже не должен будет ничего знать о пропсах базового компонента

agadoev commented on src/components/link/Link.ts:
```diff
@@ -13,6 +15,16 @@ class Link extends BaseBlock {
 
     protected componentDidMount(): void {
         this._props.class = `${this._props.class} base-link`;
+        const link = this.getContent() as HTMLLinkElement;
```
Здесь можно сделать такую же штуку, как и с querySelector. Это не обязательно, но если получится - будет очень  круто
```suggestion
        const link = this.getContent<HTMLLinkElement>();
```

agadoev commented on src/data/chats.ts:
```diff
@@ -1,36 +1,10 @@
 export type ChatData = {
+    id: number,
     name: string,
     time: string,
```
Почему бы тут не использовать тип Date? Со временем, которое представлено типом Date, гораздо удобнее работать, чем со временем, которое лежит в строке. Нам не придется мучаться и парсить его.
Возмонжо сейчас оно типа string, потому что нам нужно его выводить на экран. Если это так, то я бы создал функцию, например такую:
```
const displayedTime = (time: Date): string => {/* логика по преобразованию Date в string */}
```

agadoev commented on src/helpers/api.ts:
```diff
@@ -26,11 +27,11 @@ class Api {
     private static _instance?: Api;
 
     static getInstance(): Api {
```
хорошее решение использовать фабричный метод 👍

agadoev commented on src/helpers/api.ts:
```diff
@@ -58,7 +59,7 @@ class Api {
     }
 
-    private _request(url: string, options: AbstractMethodRequestOptions): Promise<void> {
+    private _request(url: string, options: AbstractMethodRequestOptions): Promise<any> {
```
Т.к. эта функция возвращает Promise, она является асинхронной. Лучше явно это указать, потому что явное лучше, чем не явное. Код должен быть настолько понятным, что при первом взгляде на него я должен понять, что тут происходит
```suggestion
    private async _request(url: string, options: AbstractMethodRequestOptions): Promise<any> {
```

agadoev commented on src/helpers/context.ts:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public async init (): Promise<void> {
+        await EntryService.getInstance().getInfo()
+            .then(data => {
+                Object.assign(this.profile, data);
+                this.isAuth = true;
+            })
+            .catch(() => this.isAuth = false);
```
Есть 2 подхода к использованию асинхронности - promise-like и async/await. Оба этих подхода нормально работают и имеют место быть, однако использовать их вместе не имеет смысла, потому что await нужен, чтобы зарезолвить промис и .then тоже резолвит промис. Нужно решить какой подходит будет использоваться в проекте и использовать один из них.
Почему только один из них? Потому что если в коде проекта будут использоваться разные подходы, это рано или поздно создаст путаницу + когда ты знаешь, что используется только один подход, не нужно тратить мыслительную энергию на принятие решения о том, какой подход выбрать

agadoev commented on src/helpers/router/router.ts:
```diff
@@ -0,0 +1,116 @@
+    start (defaultPathname?: string): void {
+        window.onpopstate = () => this._onRoute(window.location.pathname);
+
+        Context.getInstance().on(['isAuth'], () => this._onAuthChange());
```
Если случайно допустить ошибку в isAuth, TS ругнется? Если нет, то код недостаточно покрыт типами и стоит написать тип для событий контекста, например так
```
type TContextEvents = 'isAuth' | '' | ... и тд
```

agadoev commented on src/services/profile-service.ts:
```diff
@@ -0,0 +1,51 @@
+            data,
+        });
+    }
+    // async changeAvatar (data): Promise<boolean> {}
```
Закомментированный код лучше не держать в репозитории

agadoev commented on src/helpers/__test__/api.test.ts:
```diff
@@ -0,0 +1,129 @@
+                text: 'test',
+            };
+            const contentType = 'application/json;charset=utf-8';
+            api.post('test', {
+                headers: {
+                    'content-type': contentType,
+                },
+                data: requestData,
+            });
```
вроде api.post является асинхронным методом, но вызывается тут синхронно. Я не уверен, что это корректно

agadoev commented on src/helpers/__test__/route.test.ts:
```diff
@@ -0,0 +1,112 @@
+    route: Route,
+    getBlock: () => BaseBlock | undefined,
+} {
+    const template = Handlebars.compile('<div>{{content}}</div>');
+    let block: TestBlock | undefined = undefined;
+    class TestBlock extends BaseBlock {
+        constructor(options: BaseBlockOptions = {}) {
+            if (block) {
+                return block;
+            }
+
+            super(options);
+            block = this;
+        }
+
+        protected render(): string | DocumentFragment {
+            return template(this._props);
+        }
+    }
```
Кажется этот функционал уже есть в тестах для Block. Предлагаю сделал класс, например, TestUtils или TestHelpers, куда можно будет вынести дублирующийся функционал

agadoev commented on src/helpers/__test__/route.test.ts:
```diff
@@ -0,0 +1,112 @@
+}
+
+describe('Route', () => {
+    before(() => {
```
Тут тонкий момент. Тесты должны быть изолированы друг от друга, т.е. вызов одного теста не должен влиять на поведение другого теста.
Здесь мокается DOM - это правильно, но я не уверен, что перед запуском каждого теста состояние дома будет одинаковым. 
Я бы здесь использовал beforeEach

agadoev reviewed (approved):
Привет, Саш
Спасибо за работу. Отмечу следующие моменты
1. Круто решение использовать Context 👍
2. Так же понравилось, как ты обернул приложение в класс App, это похоже немного на C# поход и класс Program. 
3. Молодец, что написал тесты, довольно исчерпывающие (хотя их можно было писать в 4 спринте)

Я оставил несколько комментариев, которые помогут сделать код еще лучше. Т.к. код меня вполне устраивает и все задания выполнены, я приму работу, однако все же попробуй перед мержем исправить те моменты, на которые я указал. В основном они относятся к TypeScript и асинхронности.

Удачи в следующем спринте :)

alxgrents committed:
sprint_3 исправил замечания по MR

Исправил все замечания кроме замечания про TestHelpers, т.к. функционал всё же немного отличается
